packages/core/src/proofs/DataIntegrity.ts

```diff
@@ -0,0 +1,342 @@
+    )
+  }
+
+  // TODO: adding the suite context to the document interferes with existing proofs, but is currently required (see https://github.com/digitalbazaar/data-integrity/issues/19).
+  const document = ensureContext(inputDocument)
+
+  const proof = {
+    ...(id ? { id } : undefined),
+    type: PROOF_TYPE,
+    verificationMethod: signer.id,
+    cryptosuite: suite.name,
+    proofPurpose: purpose,
+  } as DataIntegrityProof
+  if (created) {
+    proof.created = created.toISOString()
+  }
+  if (expires) {
+    proof.expires = expires.toISOString()
+  }
+  if (challenge) {
+    proof.challenge = challenge
+  }
+  if (domain) {
+    proof.domain = domain
+  }
+  if (previousProof) {
+    proof.previousProof = previousProof
+  }
+
+  const canonizedProof = await suite.canonize({
+    // TODO: It's unclear if this behaviour is desirable (see https://github.com/digitalbazaar/data-integrity/issues/19).
+    // Adding the document context to the proof should NOT happen for a jcs proof according to the relevant specs;
+    // however both digitalbazaar/jsonld-signatures as well as digitalbazaar/data-integrity currently do enforce this.
+    // Not sure how to proceed; implement to spec or allow interoperability.
+    '@context': document['@context'],
+    ...proof,
+  })
+  const canonizedDoc = await suite.canonize(document)
+
+  const combinedHashes = u8aConcat(
+    sha256AsU8a(canonizedProof),
+    sha256AsU8a(canonizedDoc)
+  )
+  const signatureBytes = await signer.sign({ data: combinedHashes })
+  proof.proofValue = MULTIBASE_BASE58BTC_HEADER + base58Encode(signatureBytes)
+
+  return { ...document, proof }
+}
+
+export class UNSUPPORTED_CRYPTOSUITE_ERROR extends Error {
+  override name = 'UNSUPPORTED_CRYPTOSUITE_ERROR'
+}
+
+export class UNEXPECTED_CONTROLLER_ERROR extends Error {
+  override name = 'UNEXPECTED_CONTROLLER_ERROR'
+}
+
+export class PROOF_TRANSFORMATION_ERROR extends Error {
+  override name = 'PROOF_TRANSFORMATION_ERROR' // Specified by cryptosuite specs
+}
+
+export class MALFORMED_PROOF_ERROR extends Error {
+  override name = 'MALFORMED_PROOF_ERROR'
+  public readonly code = -17
+}
+
+export class MISMATCHED_PROOF_PURPOSE_ERROR extends Error {
+  override name = 'MISMATCHED_PROOF_PURPOSE_ERROR'
+  public readonly code = -18
+}
+
+export class INVALID_DOMAIN_ERROR extends Error {
+  override name = 'INVALID_DOMAIN_ERROR'
+  public readonly code = -19
+}
+
+export class INVALID_CHALLENGE_ERROR extends Error {
+  override name = 'INVALID_CHALLENGE_ERROR'
+  public readonly code = -20
+}
+
+export class INVALID_PROOF_PURPOSE_FOR_VERIFICATION_METHOD extends Error {
+  override name = 'INVALID_PROOF_PURPOSE_FOR_VERIFICATION_METHOD'
+  public readonly code = -25
+}
+
+async function retrieveVerificationMethod(
+  proof: DataIntegrityProof
+): Promise<VerificationMethod> {
+  const { did } = parse(proof.verificationMethod as DidUrl)
+  const { didDocument, didDocumentMetadata } = await resolve(did)
```

Comment:
For the type we can just use whatever the resolver returns. Or what do you mean?

Reply:
FOr the check true, but I said that since you use `didDocument?` everywhere, as Typescript does not know whether that exists or not. I suggested to checek for its existence as to be able to use `didDocument` directly.

Reply:
> For the type we can just use whatever the resolver returns. Or what do you mean?

The resolver returns a string, that's why I'm asking

Reply:
Do you think `SDKErrors.DidError(message)` would suffice?

Reply:
24c11745da83fdaca9a7bae229c0b9c7cca51a19